clj/src/com/sixsq/slipstream/clj_client/run.clj

```diff
@@ -1,238 +1,257 @@
+              (sort))))
 
-(defn get-node-ids
-  "IDs of the node as list."
-  [node-name]
-  (->> (-> (try+
-             (get-rtp node-name nil "ids")
-             (catch [:status 412] {}  ""))
-           (u/split  #",")
-           (sort))
-       (remove #(= (count %) 0))))
 
 ;; Predicates.
 (defn aborted?
+  "Check if run is in \"Aborted\" state."
   []
-  (not (empty? (get-abort))))
+  (not (s/blank? (get-abort))))
 
 (defn scalable?
+  "Check if run is scalable."
   []
-  (-> (http/get ru/run-url scale-req-params)
+  (-> (http/get ri/run-url ri/scale-req-params)
       :body
       xml/parse-str
       :attrs
       :mutable
       read-string))
 
-(defn- in-final-states?
-  ([]
-   (in-final-states? (get-state)))
-  ([state]
-   (u/in? state ru/non-scalable-final-states)))
-
-; TODO: use single call to get run and work on it.
 (defn can-scale?
+  "Check if it's possible to scale the run."
   []
+  ; TODO: Use single call to get run representation and query it locally.
+  ;       At the moment it's xml, which is not that comfortable to parse.
   (and
     (scalable?)
     (not (aborted?))
-    (u/in? (get-state) ru/scalable-states)))
+    (u/in? (get-state) ri/scalable-states)))
+
+(defn get-run-info
+  []
+  {:url       ri/run-url
+   :state     (get-state)
+   :scalable  (scalable?)
+   :can-scale (can-scale?)
+   :aborted   (aborted?)
+   :abort-msg (get-abort)})
+
 
 ;; Actions on the run.
+(def wait-timeout-default 600)
+
 (defn cancel-abort
-  "Cancel abort of the run.
-  "
+  "Cancel abort on the run."
   []
   (http/delete
-    ru/run-abort-url
-    rtp-req-params))
+    ri/run-abort-url
+    ri/param-req-params))
 
 (defn terminate
-  "Synchronous termination.
-  "
+  "Terminate the run."
   []
   (http/delete
-    ru/run-url
-    ru/base-req-params))
+    ri/run-url
+    ri/base-http-params))
 
 (defn scale-up
-  "Scale up node-name by n. Allow to set rtps RTPs on the new node instances.
-  Returns list of added node instance names qualified with IDs.
-  "
-  ([node-name n]
+  "Scale up application component 'comp' by 'n' instances. Allow to set parameters
+  'params' on the new component instances. Returns list of added component
+  instance names qualified with IDs."
+  ([comp n]
    (u/split (:body (http/post
-                     (ru/build-node-url node-name)
-                     (merge scale-req-params {:body (str "n=" n)})))
+                     (ri/build-component-url comp)
+                     (merge ri/scale-req-params {:body (str "n=" n)})))
             #","))
-  ([node-name n rtps]
-    (let [added-node-instances (scale-up node-name n)]
-      (doseq [[k v] rtps]
-        (doseq [id (ru/extract-ids added-node-instances)]
-          (do
-            (println "Setting RTP" node-name id k v)
-            (set-rtp node-name id k v))))
-      added-node-instances)))
+  ([name n params]
+   (let [added-instances (scale-up name n)]
+     (doseq [[k v] params]
+       (doseq [id (ri/extract-ids added-instances)]
```

Comment:
Presumably the list of ids is constant for this 'loop', so it would be better to extract it out into the above `let`. This `doseq` could also then be refactored as just a `map`.